node/src/components/block_proposer.rs

```diff
@@ -384,16 +361,11 @@ impl BlockProposerReady {
             trace!(%hash, "expired deploy rejected from the buffer");
             return;
         }
-        if self.unhandled_finalized.remove(hash.deploy_hash()) {
-            info!(%hash, "deploy was previously marked as finalized, storing header");
-            self.sets
-                .finalized_deploys
-                .insert(hash.into(), deploy_info.header);
-            return;
-        }
         // only add the deploy if it isn't contained in a finalized block
         if self.sets.finalized_deploys.contains_key(hash.deploy_hash()) {
-            info!(%hash, "deploy rejected from the buffer");
+            info!(%hash, "finalized deploy rejected from the buffer");
+            self.sets
+                .add_finalized(*hash.deploy_hash(), deploy_info.header.expires());
```

Comment:
Seems weird that `add_deploy` adds a deploy to the _finalized_ collection.

Reply:
It's only if the deploy is already in that collection.
We add it again to update its expiry date: We probably added it earlier because we learned about the finalized block before we learned about the deploy itself, so we inserted it with the maximum possible expiry date (_block_ timestamp plus _maximum_ deploy TTL). Now we can replace it with the actual expiry date (_deploy_ timestamp plus _actual_ deploy TTL), so we'll be able to purge it from the collection earlier.

Reply:
Yes, I understand it. I just wanted to note that it seems weird that we had known about deploy being finalized but didn't have the exact expiry date (b/c we based our knowledge only on the block's timestamp even though we _must_ have had the deploy already - since we can't finalize without downloading all dependencies) and now update it incidentally via `add_deploy` method which should be used for different purposes - like adding _new_ deploys to the block proposer to include in the future.

Reply:
Yes, I think it's just for the edge case where one event arrives before the other. I'm not sure it can actually happen, but it seems safer to avoid relying on even order?